CombinationSum.java

```diff
@@ -0,0 +1,56 @@
+import java.util.ArrayList;
+
+/**
+ * Given N and S, print all combinations of N variables such that their sum equals to S
+ *   Example 1: N = 2, S = 5; Output: (0,5  1,4  2,3  3,2  4,1  5,0)
+ *   Example 2: N = 3, S = 5; Output: (0,0,5  0,1,4  …  1,0,4  1,1,3  …)
+ */
+public class CombinationSum {
+    public static void main(String[] args) {
+        test(0, 0,0);
+        test(1, 5,1);
+        test(2, 5, 2);
+        test(3, 8, 3);
+        test(4, 5, 4);
+        test(5, 5, 5);
+    }
+    
+    private static void test(int N, int S, int seq) {
+        System.out.printf("[N: %d, S: %d] = {", N, S);
+        String[] prefix = new String[1];
+        prefix[0] = "";
+        ArrayList<String> result = new ArrayList<>();
+        findCombination(N, S, prefix, result, seq);
+        printList(result);
+        System.out.println("}\n");
+    }
+    
+    private static void findCombination(int N, int S, String[] prefix, ArrayList<String> result, int seq) {
+        if (N == 0) {
+            return;
+        }
+        
+        if (N == 1) {
+            prefix[0] += S;
+            result.add(prefix[0]);
```

Comment:
You can just print here, instead of putting it in an array and then printing the whole array. The array will become too big, since for some values of N and S, there are many valid combinations.